File: goblin-dao-mongo/src/main/java/org/goblinframework/dao/mongo/bson/deserializer/BsonDateDeserializer.java
package org.goblinframework.dao.mongo.bson.deserializer;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import org.goblinframework.core.util.DateUtils;

import java.io.IOException;
import java.util.Date;

final public class BsonDateDeserializer extends JsonDeserializer<Date> {

  @Override
  public Date deserialize(JsonParser p, DeserializationContext ctx) throws IOException {
    Object value = p.getEmbeddedObject();
    return DateUtils.parseDate(value);
  }
}

File: goblinframework-database-mongo/src/main/java/org/goblinframework/database/mongo/bson/deserializer/BsonDateDeserializer.kt
package org.goblinframework.database.mongo.bson.deserializer

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import org.goblinframework.core.util.DateFormatUtils
import java.util.*

class BsonDateDeserializer : JsonDeserializer<Date>() {

  override fun deserialize(p: JsonParser, ctxt: DeserializationContext?): Date? {
    return parseDate(p.embeddedObject)
  }

  companion object {
    fun parseDate(`object`: Any?): Date? {
      if (`object` is Date) {
        return `object`
      }
      if (`object` is Int || `object` is Long || `object` is Double) {
        val number = `object` as Number
        return Date(number.toLong())
      }
      return if (`object` is String) {
        DateFormatUtils.parse(`object`)
      } else null
    }
  }
}
